build/pause/pause.c

```diff
@@ -1,36 +1,175 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
 
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
 */
 
+/*
+ * From    : https://github.com/Yelp/dumb-init
+ * version : v1.1.2
+ * SHA     : eb86ebe961cbe3c2779672cd698b19f19838c5b1
+ */
+
+
+#include <assert.h>
+#include <errno.h>
+#include <getopt.h>
 #include <signal.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/wait.h>
 #include <unistd.h>
+#include <dirent.h>
+#include <ctype.h>
+
+#define PRINTERR(...) do { \
+    fprintf(stderr, "[pause] " __VA_ARGS__); \
+} while (0)
 
-static void sigdown(int signo) {
-	psignal(signo, "shutting down, got signal");
-	exit(0);
+#define DEBUG(...) do { \
+    if (debug) { \
+        PRINTERR(__VA_ARGS__); \
+    } \
+} while (0)
+
+// Signals we care about are numbered from 1 to 31, inclusive.
+// (32 and above are real-time signals.)
+#define MAXSIG 31
+
+// Indices are one-indexed (signal 1 is at index 1). Index zero is unused.
+int signal_rewrite[MAXSIG + 1] = {[0 ... MAXSIG] = -1};
+
+char debug = 0;
+
+int translate_signal(int signum) {
+    if (signum <= 0 || signum > MAXSIG) {
+        return signum;
+    } else {
+        int translated = signal_rewrite[signum];
+        if (translated == -1) {
+            return signum;
+        } else {
+            DEBUG("Translating signal %d to %d.\n", signum, translated);
+            return translated;
+        }
+    }
 }
 
-int main() {
-	if (signal(SIGINT, sigdown) == SIG_ERR)
-		return 1;
-	if (signal(SIGTERM, sigdown) == SIG_ERR)
-		return 2;
-	signal(SIGKILL, sigdown);
-	for (;;) pause();
-	fprintf(stderr, "error: infinite loop terminated\n");
-	return 42;
+void forward_signal(int signum) {
+    signum = translate_signal(signum);
+    if (signum != -1) {
+        DIR *dp = opendir ("/proc");
```

Comment:
This is really racy.  to do this right you would keep a list of all the pids you already signalled, and loop to see if there are any new ones and repeat until it come back "no more".

Couldn't this also go through the cgroup interface?  Read /proc/self/cgroups, pick one as the "canonical" group and read the tasks file.  I know lmctfy handled this well, but I have lost some state since then.

I guess if this is forked code, we just leave it alone, but this is sketchy as heck.